<a href="https://colab.research.google.com/github/myacarrizosa/Airbnb-Rating-Prediction/blob/main/04_Neural_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Neural Networks

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.preprocessing import StandardScaler

from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.regularizers import l2
from tensorflow.keras.layers import Dropout
from tensorflow.keras.callbacks import EarlyStopping

from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay

import random
random.seed(222)

In [ ]:
X_train = pd.read_csv("https://media.githubusercontent.com/media/myacarrizosa/Airbnb-Rating-Prediction/main/data/X_train_r.csv")
X_test = pd.read_csv("https://media.githubusercontent.com/media/myacarrizosa/Airbnb-Rating-Prediction/main/data/X_test_r.csv")
y_train = pd.read_csv("https://media.githubusercontent.com/media/myacarrizosa/Airbnb-Rating-Prediction/main/data/y_train_r.csv")
y_test = pd.read_csv("https://media.githubusercontent.com/media/myacarrizosa/Airbnb-Rating-Prediction/main/data/y_test_r.csv")

In [ ]:
X_train = X_train['lem_comments']
X_test = X_test['lem_comments']

y_train = y_train['type']
y_test = y_test['type']

In [ ]:
nn_y_train = y_train.values
nn_y_test = y_test.values

In [ ]:
num_words = 2000
oov_token = '<UNK>'
pad_type = 'post'
trunc_type = 'post'

In [ ]:
tokenizer = Tokenizer(num_words=num_words, oov_token=oov_token)
tokenizer.fit_on_texts(X_train)

In [ ]:
train_sequences = tokenizer.texts_to_sequences(X_train)
maxlen = max([len(x) for x in train_sequences])
train_padded = pad_sequences(train_sequences, padding=pad_type, truncating=trunc_type, maxlen=maxlen)

In [ ]:
test_sequences = tokenizer.texts_to_sequences(X_test)
test_padded = pad_sequences(test_sequences, padding=pad_type, truncating=trunc_type, maxlen=maxlen)

In [ ]:
train_padded.shape, test_padded.shape

((24510, 1004), (8171, 1004))

We ran our text data through the tensorflow keras tokenizer tool to get it processed to be run through the neural network models.

Code adapted from: https://www.kdnuggets.com/2020/03/tensorflow-keras-tokenization-text-data-prep.html

In [ ]:
ss = StandardScaler()
X_train_sc = ss.fit_transform(train_padded)
X_test_sc = ss.transform(test_padded)

In [ ]:
X_train_sc.shape, X_test_sc.shape

((24510, 1004), (8171, 1004))

We then standardized our data 

### **MODEL 1**

In [ ]:
model1 = Sequential()

model1.add(Dense(64, activation = 'relu',
                input_shape = (1004,)))
model1.add(Dense(32, activation = 'relu'))

model1.add(Dense(1, activation = 'sigmoid'))

model1.compile(loss = 'bce', optimizer = 'adam', metrics = ['accuracy'])


res1 = model1.fit(X_train_sc, nn_y_train,
                epochs = 100,
                batch_size = 64,
                validation_data = (X_test_sc, nn_y_test),
                verbose = 1) 

Epoch 1/100
383/383 [==============================] - 3s 5ms/step - loss: 0.7047 - accuracy: 0.5454 - val_loss: 1.3047 - val_accuracy: 0.5507
Epoch 2/100
383/383 [==============================] - 2s 4ms/step - loss: 0.6721 - accuracy: 0.5792 - val_loss: 1.5677 - val_accuracy: 0.5624
Epoch 3/100
383/383 [==============================] - 4s 9ms/step - loss: 0.6588 - accuracy: 0.5956 - val_loss: 1.4018 - val_accuracy: 0.5490
Epoch 4/100
383/383 [==============================] - 4s 9ms/step - loss: 0.6458 - accuracy: 0.6096 - val_loss: 1.7296 - val_accuracy: 0.5501
Epoch 5/100
383/383 [==============================] - 3s 9ms/step - loss: 0.6326 - accuracy: 0.6222 - val_loss: 2.3654 - val_accuracy: 0.5513
Epoch 6/100
383/383 [==============================] - 3s 8ms/step - loss: 0.6189 - accuracy: 0.6317 - val_loss: 2.8521 - val_accuracy: 0.5532
Epoch 7/100
383/383 [==============================] - 3s 8ms/step - loss: 0.6049 - accuracy: 0.6451 - val_loss: 2.9312 - val_accuracy: 0.5447

In [ ]:
print('Train Accuracy:')
print(res1.history['accuracy'][-1])
print('Test Accuracy:')
print(res1.history['val_accuracy'][-1])

Train Accuracy:
0.8103223443031311
Test Accuracy:
0.5327377319335938


Our first neural net model consisted of three dense layers, ending in a sigmoid activation for binary classification. We fit our model using an epoch size of 100 and a batch size of 64. Our model had a train accuracy of 0.81 and a test accuracy of 0.53, meaning that the model was very overfitted.



# Model 2 
We added more layers to try to improve our model performance 

In [ ]:
model2 = Sequential()

model2.add(Dense(64, activation = 'relu',
                input_shape = (1004,)))
model2.add(Dense(64, activation = 'relu'))
model2.add(Dense(64, activation = 'relu'))
model2.add(Dense(32, activation = 'relu'))
model2.add(Dense(32, activation = 'relu'))

model2.add(Dense(1, activation = 'sigmoid'))

model2.compile(loss = 'bce', optimizer = 'adam', metrics = ['accuracy'])


res2 = model2.fit(X_train_sc, nn_y_train,
                epochs = 100,
                batch_size = 64,
                validation_data = (X_test_sc, nn_y_test),
                verbose = 1) 

Epoch 1/100
383/383 [==============================] - 3s 6ms/step - loss: 0.6881 - accuracy: 0.5443 - val_loss: 0.7244 - val_accuracy: 0.5616
Epoch 2/100
383/383 [==============================] - 2s 5ms/step - loss: 0.6808 - accuracy: 0.5657 - val_loss: 0.7772 - val_accuracy: 0.5495
Epoch 3/100
383/383 [==============================] - 2s 5ms/step - loss: 0.6742 - accuracy: 0.5726 - val_loss: 0.8689 - val_accuracy: 0.5564
Epoch 4/100
383/383 [==============================] - 2s 6ms/step - loss: 0.6629 - accuracy: 0.5916 - val_loss: 1.2576 - val_accuracy: 0.5429
Epoch 5/100
383/383 [==============================] - 2s 5ms/step - loss: 0.6476 - accuracy: 0.6077 - val_loss: 1.6027 - val_accuracy: 0.5604
Epoch 6/100
383/383 [==============================] - 2s 5ms/step - loss: 0.6341 - accuracy: 0.6205 - val_loss: 2.3725 - val_accuracy: 0.5534
Epoch 7/100
383/383 [==============================] - 2s 5ms/step - loss: 0.6112 - accuracy: 0.6369 - val_loss: 3.4823 - val_accuracy: 0.5531

In [ ]:
print('Train Accuracy:')
print(res2.history['accuracy'][-1])
print('Test Accuracy:')
print(res2.history['val_accuracy'][-1])

Train Accuracy:
0.8437780737876892
Test Accuracy:
0.5362868905067444


Our second neural net model consisted of six dense layers, ending in a sigmoid activation for binary classification. We added these extra layers to try to improve the accuracy of the model by making it more complex. We fit our model using an epoch size of 100 and a batch size of 64. Our model had a train accuracy of 0.84 and a test accuracy of 0.54, meaning that the model was very overfitted. 

## Model 3

We then tried regularize the model by adding early stopping, kernal regularizer, and dropout layers

In [ ]:
# Build model using early stopping
es = EarlyStopping(patience = 5)

In [ ]:
model3 = Sequential()

model3.add(Dense(64, activation = 'relu',
                input_shape = (1004,)))
model3.add(Dense(64, activation = 'relu'))
model3.add(Dense(32, activation = 'relu', kernel_regularizer = l2(0.005)))
model3.add(Dropout(.5))
model3.add(Dense(32, activation = 'relu', kernel_regularizer = l2(0.005)))
model3.add(Dense(64, activation = 'relu'))
model3.add(Dense(32, activation = 'relu'))
model3.add(Dropout(.5))
model3.add(Dense(32, activation = 'relu'))
model3.add(Dense(32, activation = 'relu'))

model3.add(Dense(1, activation = 'sigmoid'))

model3.compile(loss = 'bce', optimizer = 'adam', metrics = ['accuracy'])


res3 = model3.fit(X_train_sc, nn_y_train,
                epochs = 100,
                batch_size = 64,
                validation_data = (X_test_sc, nn_y_test),
                callbacks = [es],
                verbose = 1) 

Epoch 1/100
383/383 [==============================] - 3s 6ms/step - loss: 0.7964 - accuracy: 0.5284 - val_loss: 0.7181 - val_accuracy: 0.5611
Epoch 2/100
383/383 [==============================] - 2s 6ms/step - loss: 0.6944 - accuracy: 0.5518 - val_loss: 0.7242 - val_accuracy: 0.5615
Epoch 3/100
383/383 [==============================] - 2s 6ms/step - loss: 0.6874 - accuracy: 0.5583 - val_loss: 0.7413 - val_accuracy: 0.5606
Epoch 4/100
383/383 [==============================] - 2s 6ms/step - loss: 0.6845 - accuracy: 0.5650 - val_loss: 0.7570 - val_accuracy: 0.5499
Epoch 5/100
383/383 [==============================] - 2s 6ms/step - loss: 0.6807 - accuracy: 0.5782 - val_loss: 0.7846 - val_accuracy: 0.5551
Epoch 6/100
383/383 [==============================] - 2s 6ms/step - loss: 0.6764 - accuracy: 0.5801 - val_loss: 1.0752 - val_accuracy: 0.5560


In [ ]:
print('Train Accuracy:')
print(res3.history['accuracy'][-1])
print('Test Accuracy:')
print(res3.history['val_accuracy'][-1])

Train Accuracy:
0.5801305770874023
Test Accuracy:
0.5559906959533691


Code for neural net models adapted from General Assembly Data Science Immersive Lessons 9.02 and 9.04

Our third neural net model consisted of nine dense layers, ending in a sigmoid activation for binary classification. We added these extra layers to try to improve the accuracy of the model by making it more complex. We also tried to address the overfitting of our first models by using regularization techniques such as dropout layers, kernel regularizers, and early stopping.
We fit our model using an epoch size of 100 and a batch size of 64. Our model had a train accuracy of 0.58 and a test accuracy of 0.56. This model did not have as much variance bias as the first two models but is still performing barely above baseline accuracy. Because some of our other models that weren't neural networks were performing substantially better than these models, we decided that neural networks may not be the best fit for our data. 